### PR TITLE
handheld-daemon: 1.1.0 -> 2.6.4

### PR DESCRIPTION
**Pull request by toast003**
## Description of changes

Lots of changes since this is a pretty big update, but some of the bigger ones are support for more devices, support for tdp settings and gamescope overlay (using packages not in nixpkgs yet)  as well as switching to the gpl v3 from the mit license.

https://github.com/hhd-dev/hhd/compare/v1.1.0...v2.6.2

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by toast003** [comment]
Something that I have noticed that doesn't work is the controller hiding udev rules.
They run chmod 000 on the original controller so that apps only see the controller managed by hhd, however that fails for some reason. If you manually run chmod on the devices, they get hidden.

Here's an example of a udev rule generated by hhd:

```
# Hides device gamepad devices stemming from input33
# Managed by HHD, this file will be autoremoved during configuration changes.
SUBSYSTEMS=="input", KERNELS=="input33", ATTRS{id/vendor}=="045e", ATTRS{id/product}=="028e", GOTO="hhd_valid"
GOTO="hhd_end"
LABEL="hhd_valid"
KERNEL=="js[0-9]*|event[0-9]*", SUBSYSTEM=="input", MODE="000", GROUP="root", TAG-="uaccess", RUN+="/bin/chmod 000 /dev/input/%k"
LABEL="hhd_end"

```

**Comment by toast003** [comment]
Didn't remember that /bin is empty in NixOS :sweat_smile:

**Comment by appsforartists** [comment]
Thanks for doing this!

I've fallen behind on the update train in large part because I know newer versions of handheld-daemon have iio rules that I haven't gotten around to testing.  (Also, I've just been busy with life stuff and not doing much gaming/tinkering.)  [Here was the first blush at making them work.](https://github.com/NixOS/nixpkgs/pull/284928)

What device(s) have you tested your upgrade on?

Please also include release notes, as linked in the PR template.

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [inline pkgs/by-name/ha/handheld-daemon/package.nix:8]: [`chmod` is already provided by `toybox`](https://landley.net/toybox/help.html).

**Comment by pbek** [comment]
I've added the service with

```nix
  services.handheld-daemon = {
    enable = true;
    user = username;
    package = (pkgs.callPackage ../../apps/handheld-daemon/package.nix { }).override {
    };
  };
```

And the daemon runs...
But what should happen now, or how could I test it (on my ROG Ally)? 😊

**Comment by pbek** [comment]
https://hhd.dev/ can connect to the daemon.

**Comment by toast003** [comment]
> I've fallen behind on the update train in large part because I know newer versions of handheld-daemon have iio rules that I haven't gotten around to testing. (Also, I've just been busy with life stuff and not doing much gaming/tinkering.) [Here was the first blush at making them work.](https://github.com/NixOS/nixpkgs/pull/284928)

Haven't really tested iio anything since my device doesn't use it as far as I know.

> What device(s) have you tested your upgrade on?

A 2023 gpd win max 2 with the 7840u 

> Please also include release notes, as linked in the PR template.
Handheld daemon is already in the release notes?

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [inline pkgs/by-name/ha/handheld-daemon/package.nix:44]: Should be updated to use toybox.

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [inline pkgs/by-name/ha/handheld-daemon/package.nix:30]: Why xlib?

**Review by appsforartists — CHANGES_REQUESTED** [review]

**Comment by appsforartists** [comment]
> Isn't handheld daemon is already in the release notes?

I guess it's only for major/breaking changes… 🤷‍♂️ 

Sorry, I'm new to this too…

**Comment by appsforartists** [comment]
@pbek On a Legion Go, the most obvious signs that it's working are that the LED lights turn blue (the default PlayStation color) and both the top face buttons (Legion) and bottom face buttons (start+select) work.

I don't know enough about the ROG Ally to give similar advice, but if you open your Steam controller settings and see a DualShock, that's a good sign it's working.

Are you on `unstable` or testing this PR specifically?

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (toast003) [reply pkgs/by-name/ha/handheld-daemon/package.nix:30]: Xlib seems to be a dependency now:
image

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (toast003) [reply pkgs/by-name/ha/handheld-daemon/package.nix:44]: Thanks to pointing that out, didn't realized it :sweat_smile: 

Btw are we using toolbox cause it's preferred by nixpkgs or cause it's already there? (or both idk)

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [inline pkgs/by-name/ha/handheld-daemon/package.nix:38]: Can you please move the comment above `postPatch` into this block, so it's at the same level as your new comment?

**Comment by pbek** [comment]
> @pbek On a Legion Go, the most obvious signs that it's working are that the LED lights turn blue (the default PlayStation color) and both the top face buttons (Legion) and bottom face buttons (start+select) work.

Hm, did LEDs didn't do anything and the start and select buttons already did work without hhd in steam and games run by steam. 🤔

> 
> I don't know enough about the ROG Ally to give similar advice, but if you open your Steam controller settings and see a DualShock, that's a good sign it's working.

I can still see the XBOX 360 controller, like without hhd.

> 
> Are you on `unstable` or testing this PR specifically?

Unstable, I just added the `package.nix` of this PR to my config, and I am using it in the service.

https://github.com/pbek/nixcfg/blob/516955f501644e14c7d0486b5bd1e090b9915c33/flake.nix#L157-L169

I was interested in making use of the additional two front facing buttons of the ROG Ally and something like brightness control...

**Comment by toast003** [comment]
> > I don't know enough about the ROG Ally to give similar advice, but if you open your Steam controller settings and see a DualShock, that's a good sign it's working.
> 
> I can still see the XBOX 360 controller, like without hhd.

Have you set controller emulation to dualsense in hhd.dev?

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (toast003) [reply pkgs/by-name/ha/handheld-daemon/package.nix:39]: Done

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [reply pkgs/by-name/ha/handheld-daemon/package.nix]: IDK if it matters.  Just suggesting to use it here because it was already in there.

(I think I chose it originally because it's a small dependency and I'd used it before to access POSIX commands.)

**Comment by antheas** [comment]
Toast's device activates the hrtrigger and creates the proper dir so that has been tested in fact

**Comment by antheas** [comment]
ROG Ally also uses the hrtrigger, and there will be appropriate logs showing if it works

**Comment by appsforartists** [comment]
@toast003 I don't have scope to test this right now.  If you tell me the current PR works as expected, I'm happy to merge this for you.

Also feel free to make yourself a maintainer of this lib if you so choose.

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [reply pkgs/by-name/ha/handheld-daemon/package.nix:39]: Ahh no, I meant the other way. 🙃 

Leave your comment where it is and move my comment inside the block:


```nix
postPatch = ''
  # handheld-daemon contains a fork of the python module `hid`, so this hook
  # is borrowed from the `hid` derivation.
  hidapi=${ hidapi }/lib/
  test -d $hidapi || { echo "ERROR: $hidapi doesn't exist, please update/fix this build expression."; exit 1; }
  sed -i -e "s|libhidapi|$hidapi/libhidapi|" src/hhd/controller/lib/hid.py

  # The generated udev rules point to /bin/chroot, which does not exist in NixOS
  chmod=${ toybox }/bin/chmod
  sed -i -e "s|/bin/chmod|$chmod|" src/hhd/controller/lib/hide.py
'';
```

**Comment by antheas** [comment]
@appsforartists the color of the leds no longer changes due to the playstation driver, the top buttons of the ally are start select

@pbek hhd fixes the bottom buttons of the ally to be guide and xbox + a in legacy mode (which opens the side menu in steam). However in the new version if you launch gamescope, it will detect that and connect to the X11 socket and send the command to open the side menu directly.

As for controlling the leds, if steam launches in the Deck UI, it will take control of them and change their colors, also allow setting them in the settings. In the legion go, the leds persist across reboots, in other devices they reset every boot. So if steam is not running you lose LED control for now.

A UI for the LEDs is planned.

**Comment by pbek** [comment]
> Have you set controller emulation to dualsense in hhd.dev?

I have

> ROG Ally also uses the hrtrigger, and there will be appropriate logs showing if it works

I can't see any of those logs with ` sudo journalctl -b -e`.

I can only see lots of those:

```
Apr 15 20:17:16 ally2 hhd[12787]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 15: warning: setlocale: LC_COLLATE: cannot change locale (de_AT.UTF-8): No such file or directory
Apr 15 20:17:16 ally2 hhd[12787]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 16: warning: setlocale: LC_CTYPE: cannot change locale (en_US.UTF-8): No such file or directory
Apr 15 20:17:16 ally2 hhd[12787]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 21: warning: setlocale: LC_NUMERIC: cannot change locale (de_AT.UTF-8): No such file or directory
Apr 15 20:17:16 ally2 hhd[12787]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 24: warning: setlocale: LC_TIME: cannot change locale (de_AT.UTF-8): No such file or directory
Apr 15 20:17:18 ally2 hhd[12794]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 15: warning: setlocale: LC_COLLATE: cannot change locale (de_AT.UTF-8): No such file or directory
Apr 15 20:17:18 ally2 hhd[12794]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 16: warning: setlocale: LC_CTYPE: cannot change locale (en_US.UTF-8): No such file or directory
```

**Comment by antheas** [comment]
Before merging, please bump the version that fixes this bug that was introduced by localization

**Comment by pbek** [comment]
> hhd fixes the bottom buttons of the ally to be guide

Ah, I see. Thank you!

**Comment by toast003** [comment]
> > Have you set controller emulation to dualsense in hhd.dev?
> 
> I have
> 
> > ROG Ally also uses the hrtrigger, and there will be appropriate logs showing if it works
> 
> I can't see any of those logs with ` sudo journalctl -b -e`.
> 
> I can only see lots of those:
> 
> ```
> Apr 15 20:17:16 ally2 hhd[12787]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 15: warning: setlocale: LC_COLLATE: cannot change locale (de_AT.UTF-8): No such file or directory
> Apr 15 20:17:16 ally2 hhd[12787]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 16: warning: setlocale: LC_CTYPE: cannot change locale (en_US.UTF-8): No such file or directory
> Apr 15 20:17:16 ally2 hhd[12787]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 21: warning: setlocale: LC_NUMERIC: cannot change locale (de_AT.UTF-8): No such file or directory
> Apr 15 20:17:16 ally2 hhd[12787]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 24: warning: setlocale: LC_TIME: cannot change locale (de_AT.UTF-8): No such file or directory
> Apr 15 20:17:18 ally2 hhd[12794]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 15: warning: setlocale: LC_COLLATE: cannot change locale (de_AT.UTF-8): No such file or directory
> Apr 15 20:17:18 ally2 hhd[12794]: /nix/store/5gldb5b71w2qk073xcvkpbzliwdwv29p-set-environment: line 16: warning: setlocale: LC_CTYPE: cannot change locale (en_US.UTF-8): No such file or directory
> ```

I have this too, however I see it too on other programs too (ssh when running a command iirc) so I assumed it wasn't an hhd issue

**Comment by antheas** [comment]
The logs are generated by this function:
https://github.com/hhd-dev/hhd/blob/79d8414f7c1a6a1f2e1deaec70d6dff5449298fe/src/hhd/http/i18n.py#L12-L33

I thought it wouldnt cause issues so I called it every time the api would update (every 2 seconds). It turns out both the login shell, and as it seems the `locale` command generate logs in some machines, crowding journalctl. So I only made it to apply on startup.

So it will still show the error, but only once.

**Comment by toast003** [comment]
Yeah @antheas 2.6.4 fixes the log spam

**Comment by toast003** [comment]
> @toast003 I don't have scope to test this right now. If you tell me the current PR works as expected, I'm happy to merge this for you after you update that comment edit.

I can't test gyro since it gets upset after suspending, but everything else seems to work well

> Also feel free to make yourself a maintainer of this lib if you so choose.

Do I do it on a separate pr or is it ok to do it here?

**Comment by antheas** [comment]
> I can't test gyro since it gets upset after suspending, but everything else seems to work well

Somebody needs to look into that. GPD 4 has the same issue. Other devices (such as the Ally) do not. Maybe its bmi260 related I do not know. A combination of hrtrigger and that. We know that doing `rmmod` before suspend and `modprobe` after fixes it.

**Comment by toast003** [comment]
If that's the case we could add that into nixos-hardware, but that's a topic for another time

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [reply pkgs/by-name/ha/handheld-daemon/package.nix:39]: Did this change get pushed.  I'm still seeing both comments smooshed above the hook.

**Comment by appsforartists** [comment]
> Do I do it on a separate pr or is it ok to do it here?

If you're already a nix maintainer, I'm happy to accept the change in this PR.  If you're not, you might need a separate one to get added to the maintainers list.

**Comment by toast003** [comment]
I am not in the maintainers list so I'll go and do it on a separate pr

**Comment by toast003** [comment]
This pr should be ready tho

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (toast003) [reply pkgs/by-name/ha/handheld-daemon/package.nix:39]: Oh no, I completely overlooked that, will fix now

**Comment by appsforartists** [comment]
Thanks.

Can you please push the commit with the inlined comments:

```nix
Ahh no, I meant the other way. 🙃

Leave your comment where it is and move my comment inside the block:

postPatch = ''
  # handheld-daemon contains a fork of the python module `hid`, so this hook
  # is borrowed from the `hid` derivation.
  hidapi=${ hidapi }/lib/
  test -d $hidapi || { echo "ERROR: $hidapi doesn't exist, please update/fix this build expression."; exit 1; }
  sed -i -e "s|libhidapi|$hidapi/libhidapi|" src/hhd/controller/lib/hid.py

  # The generated udev rules point to /bin/chroot, which does not exist in NixOS
  chmod=${ toybox }/bin/chmod
  sed -i -e "s|/bin/chmod|$chmod|" src/hhd/controller/lib/hide.py
'';
```

and then I'll approve?

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (toast003) [reply pkgs/by-name/ha/handheld-daemon/package.nix:38]: Ok it's pushed now

**Review by appsforartists — APPROVED** [review]

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [inline pkgs/by-name/ha/handheld-daemon/package.nix:42]: BTW, is this supposed to be `chroot` or `chmod`? 🤔

**Review by appsforartists — APPROVED** [review]

**Comment by appsforartists** [comment]
@NickCao - any chance you wanna merge this one too?

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (NickCao) [inline pkgs/by-name/ha/handheld-daemon/package.nix:38]: Please format the whole file with `nixfmt-rfc-style`

Comment (NickCao) [reply pkgs/by-name/ha/handheld-daemon/package.nix:38]: > Is nixpkgs-fmt not developed anymore?

Maybe not? But nixfmt has been ratified as the official formatter for nixpkgs: https://github.com/NixOS/rfcs/pull/166.

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (toast003) [reply pkgs/by-name/ha/handheld-daemon/package.nix:45]: Only for chmod or for hidapi on line 41?

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (toast003) [reply pkgs/by-name/ha/handheld-daemon/package.nix]: Got it got it

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (NickCao) [reply pkgs/by-name/ha/handheld-daemon/package.nix:45]: Both.

Comment (appsforartists) [reply pkgs/by-name/ha/handheld-daemon/package.nix:45]: cc @AndersonTorres for https://github.com/NixOS/nixpkgs/blob/787303bccb23278777342e927d3844e4f7187ee6/pkgs/development/python-modules/hid/default.nix

**Review by appsforartists — APPROVED** [review]
redoing LG because the bot insisted, but @NickCao made the requests, so he should be the one doing the reviewing.

File: pkgs/by-name/ha/handheld-daemon/package.nix
Comment (appsforartists) [inline pkgs/by-name/ha/handheld-daemon/package.nix:46]: This line is below the `substituteInPlace` now.  That seems like a mistake.

**Review by appsforartists — CHANGES_REQUESTED** [review]
Undoing LG because the update looks like it breaks the order of commands in `postPatch`.

**Comment by toast003** [comment]
Wdym @appsforartists ?

It builds just fine

**Comment by appsforartists** [comment]
As the comment states, this was originally cargo-culted from @AndersonTorres's `postPatch` in `hid`; however:

```nix
    # handheld-daemon contains a fork of the python module `hid`, so this hook
    # is borrowed from the `hid` derivation.
    substituteInPlace src/hhd/controller/lib/hid.py \
      --replace-fail libhidapi ${hidapi}/lib/libhidapi
    hidapi=${hidapi}/lib/
    test -d $hidapi || { echo "ERROR: $hidapi doesn't exist, please update/fix this build expression."; exit 1; }
```

uses `hidapi`, then defines it, then tests for it (after it's already been used).  

The `hidapi` declaration should come first (and fixed-up as necessary so you don't end up with any weirdness like `/lib/lib`).

**Comment by appsforartists** [comment]
Also, this whole refactoring thing feels like it's needlessly blocking the merge.

@NickCao, would you be amenable to merging this as it originally was with `sed`, and then @toast003 can break the `substituteInPlace` refactor into a new PR that includes both these sites, as well as the original in `hid`?

**Comment by toast003** [comment]
substituteInPlace doesn't use the hidapi bash variable that gets declared later, but I can put it on top

**Comment by toast003** [comment]
I don't mind, if that was the only issue I could do it here but whatever is best

**Comment by AndersonTorres** [comment]
Wut??

**Comment by appsforartists** [comment]
I misunderstood - I was reading that as a nix variable, not a bash one.

I do think we should change `hid` and this in sync, so the hooks continue to match, but I really don't like scope creep.

**Comment by NickCao** [comment]
> @NickCao, would you be amenable to merging this as it originally was with `sed`

Fine.

**Comment by toast003** [comment]
Ok @NickCao undid the refactor, you can merge now

**Review by NickCao — APPROVED** [review]

**Comment by Faupi** [comment]
With the "not refactored" `sed` solution, substituting doesn't actually seem to work (paths still seem to map to `/bin/chmod`, etc.), switching to the `substituteInPlace` fully resolved it - and it can even be seen on the native controller hiding properly.
So from my side the reverted refactor looks fine. Could we get an update on it? Currently the package is half-broken due to it.

**Comment by toast003** [comment]
I wanted to get https://github.com/NixOS/nixpkgs/pull/305027 ready before doing the refactor but I have been very busy lately (and will be until exams are done) so I haven't really had the energy to work on nix things :sweat_smile: .

I checked the package's results and paths are being changed, so I'm confused on why it isn't for you.

**Pull request merged**
